### PR TITLE
Dictionary support

### PR DESCRIPTION
**Pull request by MaxPeltzer**
#50

### CODE REVIEW

**Review by stuz5000 — APPROVED** [review]

**Review by stuz5000 — APPROVED** [review]
Looks good to me.

**Review by stuz5000 — APPROVED** [review]

**Review by stuz5000 — APPROVED** [review]
Looks good to me

**Comment by stuz5000** [comment]
Hmm... I reviewed the changes, reviewed the files, and hit approve. Nothing happened.

**Comment by MaxPeltzer** [comment]
Thanks for the review and your work that made this possible in the first place!
I guess the maintainer has to merge it.

File: pyswip/easy.py [hunk 1]
```diff
@@ -434,11 +434,46 @@ def getTerm(t):
...
+            if PL_get_arg(a, term, a0 + i):
+                args.append(getTerm(a0 + i))
+            else:
+                raise Exception("Missing arg")
+
+        def pairwise(t):
+            it = iter(t)
+            return zip(it, it)
+
+        d = {k.value: v for v, k in pairwise(args[1:])}
```
Comment (yuce) [inline pyswip/easy.py:469]: Can this be simplified to:
```
it = iter(args[1:])
d = {k.value: v for v, k in zip(it, it)}
```

File: tests/test_foreign.py [hunk 2]
```diff
@@ -85,6 +85,14 @@ def get_list_of_lists(result):
         result = list(prolog.query("get_list_of_lists(Result)"))
         self.assertTrue({'Result': [[1], [2]]} in result, 'Nested lists should be unified correctly as return value.')
 
+    def test_dictionary(self):
+        prolog = Prolog()
+        result = list(prolog.query("X = dict{value1:value1 , value2: value2}"))
```
Comment (yuce) [inline tests/test_foreign.py:90]: How about differentiating the key and the value?
```
X = dict{key1:value1 , ...
```

File: tests/test_foreign.py [hunk 3]
```diff
@@ -85,6 +85,14 @@ def get_list_of_lists(result):
         result = list(prolog.query("get_list_of_lists(Result)"))
         self.assertTrue({'Result': [[1], [2]]} in result, 'Nested lists should be unified correctly as return value.')
 
+    def test_dictionary(self):
```
Comment (yuce) [inline tests/test_foreign.py:88]: Are more test cases necessary? Like an empty dictionary?

File: tests/test_foreign.py [hunk 4]
```diff
@@ -85,6 +85,14 @@ def get_list_of_lists(result):
...
+        prolog = Prolog()
+        result = list(prolog.query("X = dict{value1:value1 , value2: value2}"))
+        dict = result[0]
+        print(dict)
```
Comment (yuce) [inline tests/test_foreign.py:92]: I think this line can be removed.

**Review by yuce — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution! Here are a few idea. Could you also add your name to the `CONTRIBUTORS.txt` file?

File: pyswip/easy.py (hunk 1 above)
Comment (MaxPeltzer) [reply pyswip/easy.py]: Simplified it as requested

File: tests/test_foreign.py (hunk 3 above)
Comment (MaxPeltzer) [reply tests/test_foreign.py:88]: Added test cases for an empty dictionary and a nested dictionary

File: tests/test_foreign.py (hunk 4 above)
Comment (MaxPeltzer) [reply tests/test_foreign.py]: Removed print statement, I don't know how I missed that

**Review by yuce — APPROVED** [review]
Looks good! Thanks again for your contribution.
